### PR TITLE
many: move encryption and installer from snap-boostrap to gadget

### PR DESCRIPTION
**Pull request by cmatsuoka**
The uc20 installer is now invoked from directly snapd instead of executing
snap-bootstrap. After refactoring the partitioning code, move encryption
support and the installer to gadget (with some bits ending up in secboot).

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

### CODE REVIEW

File: overlord/devicestate/handlers_install.go [hunk 1]
```diff
@@ -26,14 +26,10 @@ import (
 
 	"gopkg.in/tomb.v2"
 
-	// TODO:UC20 look into merging
-	// snap-bootstrap/bootstrap|partition into gadget or
-	// subpackages there cleanly
-	"github.com/snapcore/snapd/cmd/snap-bootstrap/bootstrap"
```
Comment (anonymouse64) [inline overlord/devicestate/handlers_install.go:32]: :tada:

File: gadget/install/encrypt.go [hunk 2]
```diff
@@ -204,7 +169,7 @@ func cryptsetupAddKey(key EncryptionKey, rkey RecoveryKey, node string) error {
 		return fmt.Errorf("cannot open recovery key pipe: %v", err)
 	}
 	n, err := file.Write(rkey[:])
-	if n != recoveryKeySize {
+	if n != len(rkey) {
```
Comment (anonymouse64) [inline gadget/install/encrypt.go:172]: previously this used a constant for the recovery key size and I see that we moved the definition for this constant to another package, secboot/encrypt.go, should that constant be exported and used here? or is it good enough that we use the length as the size of the recovery key?

File: secboot/encrypt.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	// On return, n == len(b) if and only if err == nil
+	return key, err
+}
+
+type RecoveryKey [recoveryKeySize]byte
+
+func NewRecoveryKey() (RecoveryKey, error) {
+	var key RecoveryKey
+	// rand.Read() is protected against short reads
+	_, err := rand.Read(key[:])
```
Comment (anonymouse64) [inline secboot/encrypt.go:52]: should this be using the randutil package instead? Perhaps not for this PR, but in general I thought there was a desire to only use the crypto/rand package from randutil where possible.

File: gadget/install/encrypt.go (hunk 2 above)
Comment (cmatsuoka) [reply gadget/install/encrypt.go:172]: I believe this should be safe since the key is an array of `recoveryKeySize` bytes, and not a slice.

Comment (cmatsuoka) [reply gadget/install/encrypt.go:172]: Also there's a forthcoming PR in the pipeline that replaces this entire function with a [helper provided by snapcore/secboot](https://github.com/snapcore/secboot/pull/54/files#diff-9662d8c5fc8c415f21e1ecc3f784ab73R565).

File: secboot/encrypt.go (hunk 3 above)
Comment (cmatsuoka) [reply secboot/encrypt.go:52]: That makes sense, I'll have a look at randutil.

File: gadget/install/partition.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/snapcore/snapd/gadget/internal"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	ensureNodesExist = ensureNodesExistImpl
+)
+
+// createMissingPartitions creates the partitions listed in the positioned volume
```
Comment (bboozzoo) [inline gadget/install/partition.go:39]: Must have been left behind when we did renames. `s/positioned/laid out/`

File: gadget/ondisk.go [hunk 5]
```diff
@@ -474,9 +363,9 @@ func buildPartitionList(dl *OnDiskVolume, pv *LaidOutVolume) (sfdiskInput *bytes
 	return buf, toBeCreated
 }
 
-// listCreatedPartitions returns a list of partitions created during the
+// ListCreatedPartitions returns a list of partitions created during the
 // install process.
-func listCreatedPartitions(layout *OnDiskVolume) []string {
+func ListCreatedPartitions(layout *OnDiskVolume) []string {
```
Comment (bboozzoo) [inline gadget/ondisk.go:368]: Perhaps a slightly different name `func CreatedDuringInstall() (created []string)`?

File: secboot/encrypt.go (hunk 3 above)
Comment (bboozzoo) [reply secboot/encrypt.go:52]: Does this still need to be updated?

Comment (anonymouse64) [reply secboot/encrypt.go:52]: I think it's fine for a followup IMHO

File: gadget/ondisk.go [hunk 6]
```diff
@@ -141,7 +139,7 @@ type OnDiskVolume struct {
 	Size Size
 	// sector size in bytes
 	SectorSize     Size
-	partitionTable *sfdiskPartitionTable
+	PartitionTable *sfdiskPartitionTable
```
Comment (pedronis) [inline gadget/ondisk.go:142]: not sure what to do immediately, it relates to the larger TODOs, but having an exported field using an internal type pointer is weird

File: gadget/ondisk.go [hunk 7]
```diff
@@ -474,10 +363,10 @@ func buildPartitionList(dl *OnDiskVolume, pv *LaidOutVolume) (sfdiskInput *bytes
 	return buf, toBeCreated
 }
 
-// listCreatedPartitions returns a list of partitions created during the
+// CreatedDuringInstall returns a list of partitions created during the
 // install process.
-func listCreatedPartitions(layout *OnDiskVolume) []string {
-	created := make([]string, 0, len(layout.Structure))
+func CreatedDuringInstall(layout *OnDiskVolume) (created []string) {
```
Comment (pedronis) [inline gadget/ondisk.go:368]: this can move to install I think

File: gadget/ondisk.go (hunk 6 above)
Comment (cmatsuoka) [reply gadget/ondisk.go:142]: Hmm, yes. Perhaps we could keep it non-exported and provide an UpdatePartitionTable() method in the reorganization.

File: gadget/ondisk.go (hunk 7 above)
Comment (cmatsuoka) [reply gadget/ondisk.go:368]: Will do in a follow-up.

**Pull request merged**